alchemist-loading/src/main/kotlin/it/unibo/alchemist/boundary/loader/VariablesOverrider.kt

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2010-2023, Danilo Pianini and contributors
+ * listed, for each module, in the respective subproject's build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.boundary.loader
+
+import it.unibo.alchemist.boundary.modelproviders.YamlProvider
+
+/**
+ * Use this to override a map of variables with a list of resolvable key-value pairs.
+ */
+object VariablesOverrider {
+
+    /**
+     * Apply overrides to a map of variables.
+     * @parameter map Map to be overriden
+     * @parameter overrides list of valid String yaml strings containing overrides
+     */
+    @JvmStatic
+    fun applyOverrides(map: Map<String, *>, overrides: List<String>): Map<String, *> {
+        return if (overrides.isEmpty()) {
+            map
+        } else {
+            val newMap = map.toMutableMap()
+            overrides.forEach { applyOverride(it, newMap) }
+            newMap
+        }
+    }
+
+    @JvmStatic
+    private fun applyOverride(override: String, newMap: MutableMap<String, Any?>) {
+        val overrideMap = YamlProvider.from(override)
+        overrideMap.forEach { entry ->
+            mergeInto(entry.key, entry.value, newMap)
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    @JvmStatic
+    private fun mergeInto(key: String, value: Any?, newMap: MutableMap<String, Any?>) {
+        when (value) {
+            is MutableMap<*, *> -> {
+                if (newMap[key] is MutableMap<*, *>) {
+                    val castValue = value as MutableMap<String, Any?>
```

Comment:
this is unchecked... and the name is not very clear. "Castvalue"? maybe "innerMap"?

Reply:
I agree, innerMap is much better. As for the unchecked cast, unfortunately I couldn't get around the type erasure limitations. I could not check for 

newMap[key] is MutableMap<String, *>

 as the type gets erased (I get this error at compile time): Cannot check for instance of erased type: MutableMap<String, *>. Same for the "when" block on the value.

I feel like unchecked casts to map with string keys is a neccessary evil here. I agree that this implementations make very strong assumptions, however a map with a non-string value wouldn't make sense in this context since yaml maps are string-keyed so I think that it's acceptable to fail and throw an exception here. Maybe the exception could be more explicit, something like IllegalArgumentException("cannot ovveride a non string-keyed map")?

Is there a better way to check the collection types here that I'm missing?

Reply:
Refactored to "innerMap"

Reply:
Yes, cast should be done to MutableMap<*, *>, and key/values checked and cast upon usage

Reply:
Done